python/ray/data/aggregate.py

```diff
@@ -71,17 +79,168 @@ def __init__(self):
         )
 
 
+def _null_init(init: Callable[[KeyType], AggType], k: KeyType):
+    """
+    Initialize accumulation while handling nulls.
+
+    This adds on a has_data field that the accumulator uses to track whether an
+    aggregation is empty.
+
+    Args:
+        init: The core init function.
+        k: Groupby key.
+
+    Returns:
+        Initialized accumulation that can be fed into the aggregation's
+        accumulate function.
+    """
+    a = init(k)
+    if not isinstance(a, list):
+        a = [a]
+    return a + [0]
```

Comment:
Does 0 mean "null"?

Reply:
0 means "empty accumulation", 1 means "accumulation has data"

Reply:
Make this clear in the code? Ideally we'd use a class here if it's not a performance issue, like NullableValue.

Reply:
It needs to be Arrow-compatible and needs to be coercible to the same data type as the accumulated Arrow data, which is why we went away from such sentinels.